File: JavaPatterns/src/main/java/behavioral/DependencyInjection.java
package behavioral;

/**
 * The dependency injection (DI) pattern allows to avoid hard-coded dependencies and
 * to substitute dependencies either at run-time or at compile time.
 * The pattern is a special case of inversion of control (IoC) technique.
 */

class User {
    private String name;

    public User(String name) {
        this.name = name;
    }

    public String getName() {
        return name;
    }
}

interface Repository {
    void save(User user);
}

class DatabaseRepository implements Repository {
    @Override
    public void save(User user) {
        System.out.println("Saving in database repository user " + user.getName());
    }
}

class FileRepository implements Repository {
    @Override
    public void save(User user) {
        System.out.println("Saving in file repository user " + user.getName());
    }
}

class UserService {
    private final Repository repository;

    UserService(Repository repository) {
        this.repository = repository;
    }

    void create(User user) {
        repository.save(user);
    }
}
public class DependencyInjection {

    public static void main(String [] args) {
        User user = new User("Gigi");
        UserService databaseService = new UserService(new DatabaseRepository());
        UserService fileService = new UserService(new FileRepository());

        databaseService.create(user);
        fileService.create(user);
    }
}

File: ScalaPatterns/src/main/scala/behavioral/DependencyInjection.scala
package behavioral

/**
 * In addition to composition (“HAS-A”) and inheritance (“IS-A”), Scala offers a special kind of object relation – requirement (“REQUIRES-A”),
 * expressed as self-type annotations. Self-types let us specify additional type expectations for an object, without exposing them in the inheritance hierarchy.
 * We can use self-type annotations together with traits to implement dependency injection
 *
 * @author Daniel Leon
 */
object DependencyInjection {

  case class User(name  : String)

  trait Repository {
    def save(user: User)
  }

  trait DatabaseRepository extends Repository {
    override def save(user : User) = {
      println("Saving in database repository user : " + user.name)
    }
  }

  trait FileRepository extends Repository {
    override def save(user : User) = {
      println("Saving in file repository user : " + user.name)
    }
  }

 class UserService { self: Repository => // requires Repository
    def create(user: User) {
      save(user)
    }
  }

  def main(args : Array[String]) {

    val frank = User("Frankie")
    val userService = new UserService with DatabaseRepository
    val fileService = new UserService with FileRepository

    userService.create(frank)
    fileService.create(frank)
  }
}
